src/main/java/org/jenkinsci/plugins/workflow/steps/durable_task/DurableTaskStep.java

```diff
@@ -325,13 +332,34 @@ private static synchronized ScheduledThreadPoolExecutor threadPool() {
             return false;
         }
 
-        private @CheckForNull FilePath getWorkspace() throws AbortException {
+        private @CheckForNull FilePath getWorkspace() throws IOException, InterruptedException {
             if (ws == null) {
                 ws = FilePathUtils.find(node, remote);
                 if (ws == null) {
+                    // Part of JENKINS-49707: check whether an agent has been removed.
+                    // (Note that a Computer may be missing because a Node is offline,
+                    // and conversely after removing a Node its Computer may remain for a while.
+                    // Therefore we only fail here if _both_ are absent.)
+                    // ExecutorStepExecution.NodeListener will normally do this first, so this is a fallback.
+                    Jenkins j = Jenkins.getInstanceOrNull();
+                    if (ExecutorStepExecution.RemovedNodeCause.ENABLED && !node.isEmpty() && j != null && j.getNode(node) == null) {
+                        if (removedNodeDiscovered == 0) {
+                            LOGGER.fine(() -> "discovered that " + node + " has been removed");
+                            removedNodeDiscovered = System.nanoTime();
+                            return null;
+                        } else if (System.nanoTime() < removedNodeDiscovered + TimeUnit.MILLISECONDS.toNanos(ExecutorPickle.TIMEOUT_WAITING_FOR_NODE_MILLIS)) {
+                            LOGGER.fine(() -> "rediscovering that " + node + " has been removed");
+                            return null;
+                        } else {
+                            LOGGER.fine(() -> node + " has been removed for a while, assuming it is not coming back");
+                            throw new FlowInterruptedException(Result.ABORTED, new ExecutorStepExecution.RemovedNodeCause());
```

Comment:
Is `ABORTED` really the result we want in this case? My initial thought was that `FAILURE` would be more appropriate. Thoughts?

Looks like at least one reason for `ABORTED` is that over in `RemovedNodeListener` you call `BodyExecution.cancel`, and `CpsBodyExecution` always throws a `FlowInterruptedException` with result `ABORTED` in that method, so it makes sense to be consistent.

Reply:
Generally `FAILURE` is used for cases where something about the build itself broke. The typical scenario here is that the build started running, something like a spot instance got killed, and so the build could not complete due to external causes. Prior to this patch (plus downstream where appropriate) the build would just hang until manually `ABORTED`; now the abort is automatic.